amethyst_gltf/src/format/material.rs

```diff
@@ -0,0 +1,196 @@
+
+    let em_factor = material.emissive_factor();
+    prefab.emission = Some(load_texture_with_factor(
+        material.emissive_texture(),
+        [em_factor[0], em_factor[1], em_factor[2], 1.0],
+        buffers,
+        source.clone(),
+        name,
+    ).map(|(texture, _)| TexturePrefab::Data(texture))?);
+    prefab.emission_id = material.index().map(|i| i as u64 * 10 + 3);
+
+    // Can't use map/and_then because of Result returning from the load_texture function
+    prefab.normal = match material.normal_texture() {
+        Some(normal_texture) => {
+            Some(
+                load_texture(&normal_texture.texture(), buffers, source.clone(), name)
+                    .map(|data| TexturePrefab::Data(data))?,
+            )
+        }
+
+        None => None,
+    };
+    prefab.normal_id = material.index().map(|i| i as u64 * 10 + 4);
+
+    // Can't use map/and_then because of Result returning from the load_texture function
+    prefab.ambient_occlusion = match material.occlusion_texture() {
+        Some(occlusion_texture) => {
+            Some(
+                load_texture(&occlusion_texture.texture(), buffers, source.clone(), name)
+                    .map(|data| TexturePrefab::Data(data))?,
+            )
+        }
+
+        None => None,
+    };
+    prefab.ambient_occlusion_id = material.index().map(|i| i as u64 * 10 + 5);
+    prefab.transparent = if let AlphaMode::Blend = material.alpha_mode() {
+        true
+    } else {
+        false
+    };
+    if let AlphaMode::Mask = material.alpha_mode() {
+        prefab.alpha_cutoff = material.alpha_cutoff();
+    }
+
+    Ok(prefab)
+}
+
+fn deconstruct_metallic_roughness(
+    data: TextureData,
+    metallic_factor: f32,
+    roughness_factor: f32,
+) -> ((TextureData, f32), (TextureData, f32)) {
+    (
+        (
+            deconstruct_image(&data, 2, 4), // metallic from B channel
+            metallic_factor,
+        ),
+        (
+            deconstruct_image(&data, 1, 4), // roughness from G channel
+            roughness_factor,
+        ),
+    )
+}
+
+fn deconstruct_image(data: &TextureData, offset: usize, step: usize) -> TextureData {
+    use gfx::format::SurfaceType;
+    match *data {
+        TextureData::Image(ref image_data, ref metadata) => {
+            let metadata = metadata
+                .clone()
+                .with_size(image_data.raw.w as u16, image_data.raw.h as u16)
+                .with_format(SurfaceType::R8);
+            let image_data = image_data
+                .raw
+                .buf
+                .iter()
+                .dropping(offset)
+                .step(step)
+                .cloned()
+                .collect();
+            TextureData::U8(image_data, metadata)
+        }
+        TextureData::Rgba(ref color, ref metadata) => {
+            TextureData::Rgba([color[offset]; 4], metadata.clone())
+        }
+        _ => unreachable!(), // We only support color and image for textures from gltf files
```

Comment:
Is this truly unreachable?

Reply:
Yes, GLTF only support material data with either a [f32; 4] or an image.

Reply:
And those are the only variants that are produced by the material loading.